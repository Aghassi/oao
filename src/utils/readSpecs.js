// @flow

import path from 'path';
import fs from 'fs';
import { mainStory } from 'storyboard';
import listPaths from './listPaths';

const ROOT_PACKAGE = '__ROOT_PACKAGE__';

type PackageName = string;
type OaoSpecs = {
  pkgPath: string,
  specPath: string,  // including .package.json
  name: string,
  displayName: string,
  specs: Object,
};
type AllSpecs = { [key: PackageName]: OaoSpecs };

const readAllSpecs = async (srcPatterns: string): Promise<AllSpecs> => {
  const pkgPaths = await listPaths(srcPatterns);
  pkgPaths.push('.');
  const allSpecs = {};
  mainStory.info('Reading all package.json files...');
  pkgPaths.forEach((pkgPath) => {
    const pkg = readOneSpec(pkgPath);
    allSpecs[pkg.name] = pkg;
  });
  return allSpecs;
};

const readOneSpec = (pkgPath: string): OaoSpecs => {
  const pkg = {};
  pkg.pkgPath = pkgPath;
  try {
    pkg.specPath = path.resolve(process.cwd(), pkgPath, 'package.json');
    pkg.specs = JSON.parse(fs.readFileSync(pkg.specPath, 'utf8'));
  } catch (err) {
    mainStory.error(`Could not read package.json at ${pkg.specPath}`);
    throw err;
  }
  const name = pkgPath === '.' ? ROOT_PACKAGE : pkg.specs.name;
  validatePkgName(pkgPath, name);
  pkg.name = name;
  const displayName = name === ROOT_PACKAGE ? 'ROOT' : name;
  pkg.displayName = displayName;
  return pkg;
};

const validatePkgName = (pkgPath: string, name: PackageName): void => {
  if (name == null || name === '') throw new Error(`Package has no name (${pkgPath})`);
  if (pkgPath === '.') return;
  const segments = pkgPath.split('/');
  if (name[0] !== '@' && name !== segments[segments.length - 1]) {
    mainStory.error(`Package name (${name}) does not match directory name ${pkgPath}`);
    throw new Error('INVALID_DIR_NAME');
  }
};

export {
  readAllSpecs,
  readOneSpec,
  ROOT_PACKAGE,
};
