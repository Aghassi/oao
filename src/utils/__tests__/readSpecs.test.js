/* eslint-env jest */
/* eslint-disable global-require */

import { readAllSpecs, ROOT_PACKAGE } from '../readSpecs';

describe('readAllSpecs', () => {
  it('reads all package.json files as well as metadata', async () => {
    const allSpecs = await readAllSpecs('test/fixtures/packages/*');
    const rootSpecs = allSpecs[ROOT_PACKAGE];
    expect(rootSpecs.pkgPath).toBe('.');
    expect(rootSpecs.displayName).toBe('MONOREPO ROOT');
    expect(rootSpecs.specs.name).toBe('oao');
    delete allSpecs[ROOT_PACKAGE];
    Object.keys(allSpecs).forEach((name) => {
      expect(allSpecs[name].specPath).not.toBeNull();
      delete allSpecs[name].specPath;
    });
    expect(allSpecs).toMatchSnapshot();
  });

  it('accepts --src ending with slash', async () => {
    const allSpecs = await readAllSpecs('test/fixtures/packages/*/');
    delete allSpecs[ROOT_PACKAGE];
    Object.keys(allSpecs).forEach((name) => {
      delete allSpecs[name].specPath;
    });
    expect(allSpecs).toMatchSnapshot();
  });

  it('supports scoped packages', async () => {
    const allSpecs = await readAllSpecs('test/fixtures/packagesScoped/*');
    expect(allSpecs['@guigrpa/example-package'].specs.name).toEqual('@guigrpa/example-package');
    expect(allSpecs['@guigrpa/example-package-b'].specs.name).toEqual('@guigrpa/example-package-b');
  });

  it('throws on invalid directory names (for non-scoped packages)', async () => {
    try {
      await readAllSpecs('test/fixtures/packagesWrongName/*');
      throw new Error('DID_NOT_THROW');
    } catch (err) {
      if (err.message !== 'INVALID_DIR_NAME') throw err;
    }
  });
});
