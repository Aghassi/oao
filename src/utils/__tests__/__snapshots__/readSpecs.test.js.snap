// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readAllSpecs reads all package.json files as well as metadata 1`] = `
Object {
  "oao": Object {
    "name": "oao",
    "pkgPath": "test/fixtures/packages/oao",
    "specs": Object {
      "license": "MIT",
      "main": "index.js",
      "name": "oao",
      "version": "0.1.0",
    },
  },
  "oao-b": Object {
    "name": "oao-b",
    "pkgPath": "test/fixtures/packages/oao-b",
    "specs": Object {
      "dependencies": Object {
        "oao": "*",
        "timm": "1.x",
      },
      "license": "MIT",
      "main": "index.js",
      "name": "oao-b",
      "version": "0.1.0",
    },
  },
  "oao-c": Object {
    "name": "oao-c",
    "pkgPath": "test/fixtures/packages/oao-c",
    "specs": Object {
      "dependencies": Object {
        "oao": "*",
        "timm": "1.x",
      },
      "devDependencies": Object {
        "oao-b": "*",
        "xxl": "1.x",
      },
      "license": "MIT",
      "main": "index.js",
      "name": "oao-c",
      "peerDependencies": Object {
        "oao-b": "*",
      },
      "version": "0.1.0",
    },
  },
  "oao-d": Object {
    "name": "oao-d",
    "pkgPath": "test/fixtures/packages/oao-d",
    "specs": Object {
      "dependencies": Object {
        "oao": "*",
        "timm": "1.x",
      },
      "devDependencies": Object {
        "oao-b": "*",
        "oao-c": "*",
        "xxl": "1.x",
      },
      "license": "MIT",
      "main": "index.js",
      "name": "oao-d",
      "peerDependencies": Object {},
      "version": "0.1.0",
    },
  },
  "oao-priv": Object {
    "name": "oao-priv",
    "pkgPath": "test/fixtures/packages/oao-priv",
    "specs": Object {
      "dependencies": Object {
        "oao": "*",
        "timm": "1.x",
      },
      "devDependencies": Object {
        "oao-b": "*",
        "oao-c": "*",
        "xxl": "1.x",
      },
      "license": "MIT",
      "main": "index.js",
      "name": "oao-priv",
      "peerDependencies": Object {},
      "private": true,
      "version": "0.1.0",
    },
  },
}
`;
